### PR TITLE
Separate the domain layer and the database layer

### LINKED ISSUE #179: Refactor repositories
Refactor the repositories so that domain models are separated from database records appropriately. Currently, the two abstraction layers leak into each other.

### PR DESCRIPTION
**Pull request by rexor12**
Closes #179 

* Separate models and records where records were used directly in the domain layer
* Remove duplication in repositories by introducing a generic base class (RepositoryBase)
* Automatic mapping of records in queries

### CODE REVIEW

File: holobot/sdk/database/repositories/repository_base.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                    .where()
+                )
+                results = await query.compile().fetch(connection)
+
+                return tuple(
+                    self._map_record_to_model(self._map_query_result_to_record(result))
+                    for result in results
+                )
+
+    async def _get_one_by_filter(
```
Comment (rexor12) [inline holobot/sdk/database/repositories/repository_base.py:196]: `_get_by_filter`

File: holobot/sdk/database/repositories/repository_base.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+    @abstractmethod
+    def _map_record_to_model(self, record: TRecord) -> TModel:
+        ...
+
+    @abstractmethod
+    def _map_model_to_record(self, model: TModel) -> TRecord:
+        ...
+
+    async def _get_by_filter(
```
Comment (rexor12) [inline holobot/sdk/database/repositories/repository_base.py:176]: `_get_many_by_filter`

File: holobot/sdk/database/repositories/repository_base.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        self,
+        filter_builder: Callable[[WhereBuilder], ICompileableQueryPartBuilder[CompiledQuery]]
+    ) -> tuple[TModel, ...]:
+        async with self._database_manager.acquire_connection() as connection:
```
Comment (rexor12) [inline holobot/sdk/database/repositories/repository_base.py:180]: docstring

File: holobot/sdk/database/repositories/repository_base.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+                return tuple(
+                    self._map_record_to_model(self._map_query_result_to_record(result))
+                    for result in results
+                )
+
+    async def _get_one_by_filter(
+        self,
+        filter_builder: Callable[[WhereBuilder], ICompileableQueryPartBuilder[CompiledQuery]]
+    ) -> TModel | None:
+        async with self._database_manager.acquire_connection() as connection:
```
Comment (rexor12) [inline holobot/sdk/database/repositories/repository_base.py:200]: docstring

File: holobot/sdk/database/repositories/repository_base.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+                )
+
+    async def _paginate(
+        self,
+        ordering_column: str,
+        page_index: int,
+        page_size: int,
+        filter_builder: Callable[[WhereBuilder], ISupportsPagination]
+    ) -> PaginationResult[TModel]:
+        async with self._database_manager.acquire_connection() as connection:
```
Comment (rexor12) [inline holobot/sdk/database/repositories/repository_base.py:222]: docstring

File: holobot/sdk/database/repositories/entity.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+from typing import Protocol, TypeVar
+
+TIdentifier = TypeVar("TIdentifier")
+
+class Entity(Protocol[TIdentifier]):
```
Comment (rexor12) [inline holobot/sdk/database/repositories/entity.py:5]: Base class?

File: holobot/sdk/database/repositories/entity.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    # NOTE By convention, only simple primary keys are supported
+    # and they must be named as "id".
+    id: TIdentifier
+
+    # NOTE This empty initializer is required to avoid a false error
+    # in RepositoryBase that believes IEntity has no initializer.
+    def __init__(self) -> None:
```
Comment (rexor12) [inline holobot/sdk/database/repositories/entity.py:12]: Double check if this is still needed, especially with the base class. If yes, fix the comment (`IEntity` -> `Entity`).

File: holobot/sdk/database/repositories/repository_base.py [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+        the runtime resolution of generic type arguments.
+
+        :return: The type of the record.
+        :rtype: type[TRecord]
+        """
+
+    @property
+    @abstractmethod
+    def table_name(self) -> str:
+        ...
```
Comment (rexor12) [inline holobot/sdk/database/repositories/repository_base.py:64]: docstring

File: holobot/sdk/database/repositories/repository_base.py [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+
+    def __init__(
+        self,
+        database_manager: DatabaseManagerInterface
```
Comment (rexor12) [inline holobot/sdk/database/repositories/repository_base.py:68]: Use this opportunity to rename `DatabaseManagerInterface` to `IDatabaseManager`.

**Review by tooruu — COMMENTED** [review]
lgtm

File: holobot/extensions/giveaways/repositories/external_giveaway_item_repository.py [hunk 10]
```diff
@@ -77,13 +57,13 @@ async def get_metadatas(
         item_type: str,
         active_only: bool = True
     ) -> PaginationResult[ExternalGiveawayItemMetadata]:
-        async with self.__database_manager.acquire_connection() as connection:
+        async with self._database_manager.acquire_connection() as connection:
```
Comment (tooruu) [inline holobot/extensions/giveaways/repositories/external_giveaway_item_repository.py:60]: POG

**Pull request merged**